src/http/server.ts

```diff
@@ -758,32 +752,241 @@ export class Server {
+      return (param.includes(":") || param.includes("{")) &&
+        !param.includes("?");
+    }).length;
+    for (let i = 0; i < numberOfRequiredParams; i++) {
+      tmpPath = tmpPath.replace(
+        /(:[^(/]+|{[^0-9][^}]*})/, // same as REGEX_URI_MATCHES but not global
+        Server.REGEX_URI_REPLACEMENT,
+      );
+    }
+    // Replace optional path params
+    const maxOptionalParams = path.split("/").filter((param) => {
+      return param.includes("?");
+    }).length;
+    // Description for the below for loop and why we use it to create the regex for
+    // optional parameters:
+    // For each optional parameter in the path, we replace it with custom regex.
+    // Similar to how other blocks construct the `regex_path`, but in this case,
+    // it isn't as easy as a simple `replace` one-liner, due to needing to account for
+    // optional parameters (:name?), and required parameters before optional params.
+    // This is what we do to construct the `regex_path`. I haven't been able to
+    // come up with a regex that would replace all instances and work, which is
+    // why a loop is being used here, to replace the first instance of an optional
+    // parameter (to account for a possible required parameter before), and then
+    // replace the rest of the occurrences. It's slightly tricky because the path
+    // `/users/:name?/:age?/:city?` should match  `/users`.
+    for (let i = 0; i < maxOptionalParams; i++) {
+      if (i === 0) { // We need to mark the start for the first optional param
+        // The below regex is very similar to `REGEX_URI_MATCHES` but this regex isn't
+        // global, and accounts for there being a required parameter before
+        tmpPath = tmpPath.replace(
+          /\/(:[^(/]+|{[^0-9][^}]*}\?)\/?/,
+          "/?([a-zA-Z0-9]+)?/?", // A `/` being optional, as well as the param being optional, and a ending `/` being optional
+        );
+      } else { // We can now create the replace regex for the rest taking into consideration the above replace regex
+        tmpPath = tmpPath.replace(
+          /\/?(:[^(/]+|{[^0-9][^}]*}\?)\/?/,
+          "([^/]+)?/?",
+        );
       }
     }
-    return undefined;
+
+    return {
+      og_path: path,
+      regex_path: `^${tmpPath}$`,
+      // Regex is same as other blocks, but we also strip the `?`.
+      params: (path.match(Server.REGEX_URI_MATCHES) || []).map(
+        (element: string) => {
+          return element.replace(/:|{|}|\?/g, "");
+        },
+      ),
+    };
+  }
+
+  /**
+   * Has the request URL path in question been handled previously? That is, a
+   * resource was found for it and an association was stored in the CACHED
+   * resource lookup Map.
+   *
+   * @param urlPath - The request URL path in question.
+   *
+   * @returns True if handled previously; false if not.
+   */
+  protected requestUrlWasHandledPreviously(urlPath: string): boolean {
+    return this.cached_resource_lookup_table.has(urlPath);
```

Comment:
index service should handle caching

Reply:
finding: it was found that drash having its own cache was faster than the index service managing a cache